filter.go

```diff
@@ -139,28 +169,70 @@ func (d *PacketFilter) Overflow() uint64 {
 // Should call this after creating all the expected connections using NewConn, otherwise the packets
 // read will be dropped.
 func (d *PacketFilter) Start() {
-	pktReader := d.readFrom
-	if d.oobConn != nil {
-		pktReader = d.readMsgUdp
+	msgReader := d.readFrom
+	if d.ipv4Conn != nil {
+		msgReader = d.readBatch
+	} else if d.oobConn != nil {
+		msgReader = d.readMsgUdp
 	}
-	go d.loop(pktReader)
+	go d.loop(msgReader)
 }
 
-func (d *PacketFilter) readFrom() packet {
+func (d *PacketFilter) readFrom() []messageWithError {
 	buf := d.bufPool.Get().([]byte)
 	n, addr, err := d.conn.ReadFrom(buf)
 
-	return packet{
-		n:    n,
-		addr: addr,
-		err:  err,
-		buf:  buf[:n],
+	return []messageWithError{
+		{
+			Message: ipv4.Message{
+				Buffers: [][]byte{buf[:n]},
+				Addr:    addr,
+				N:       n,
+			},
+			Err: err,
+		},
 	}
 }
 
+func (d *PacketFilter) readBatch() []messageWithError {
+	batch := make([]ipv4.Message, d.batchSize)
+	for i := range batch {
+		buf := d.bufPool.Get().([]byte)
+		oobBuf := d.bufPool.Get().([]byte)
+		batch[i].Buffers = [][]byte{buf}
+		batch[i].OOB = oobBuf
+	}
+
+	n, err := d.ipv4Conn.ReadBatch(batch, 0)
+
+	// This is entirely unexpected, but happens in the wild
+	if n < 0 && err == nil {
+		err = errUnexpectedNegativeLength
+	}
+
+	if err != nil {
+		// Pretend we've read one message, so we reuse the first message of the batch for error
+		// propagation.
+		n = 1
+	}
+
+	result := make([]messageWithError, n)
+
+	for i := 0; i < n; i++ {
+		result[i].Err = err
+		result[i].Message = batch[i]
+	}
+
+	for _, msg := range batch[n:] {
+		d.returnBuffers(msg)
+	}
```

Comment:
Aren't buffers returned in outer functions (`loop`, `ReadFrom`, ...)?

Reply:
Yes, the ones that we return as part of result (`for i := 0; i < n; i++ {`), everything after n has to be returned by us, as we are not returning that back to the caller.